### PR TITLE
[add] DETR Instance Sesgmentation via Detectron2

### PR DESCRIPTION
**Pull request by jd730**
Implement DETR Instance Segmentation using Detectron2.

- Potential Refinements:
1. Import `polygons_to_bitmask` from detectron2 in `models/detr.py`
2. The naive way to load pre-trained weight. (`WEIGHTS` on did not work)
3. Manual flag for detectron2 mode on `SetCriterion`
4. Expand to Panoptic Segmentation

Related #144

### CODE REVIEW

File: models/detr.py [hunk 1]
```diff
@@ -161,6 +162,22 @@ def loss_boxes(self, outputs, targets, indices, num_boxes):
...
+            h, w = t["size"]
+            polygon = t["masks"].polygons[m[i]]
+            for p in polygon:
+                p[0::2] *= (size[1] / w)
+                p[1::2] *= (size[0] / h)
+            masks.append(polygons_to_bitmask(polygon, size[0], size[1]))
+        masks = torch.tensor(masks)
+        if len(masks.shape) != 3:
+            masks = masks.unsqueeze(0)
+        return masks
```
Comment (fmassa) [inline models/detr.py:179]: This should not live in `models/detr`, but instead in the d2 wrapper code to convert to the representation that we expect.

This way, the model in `detr` only expects a single type of input data, making things simpler overall.

File: util/misc.py [hunk 2]
```diff
@@ -429,3 +433,20 @@ class can go away.
         return _new_empty_tensor(input, output_shape)
     else:
         return torchvision.ops.misc.interpolate(input, size, scale_factor, mode, align_corners)
+
+
+# Copied from Detectron2:
+# https://github.com/facebookresearch/detectron2/blob/af866c423b5897c8ab709a405c86a45b51ad4785/detectron2/structures/masks.py#L20
+def polygons_to_bitmask(polygons: List[np.ndarray], height: int, width: int) -> np.ndarray:
```
Comment (fmassa) [inline util/misc.py:440]: This is already present in https://github.com/facebookresearch/detr/blob/9db8be19e76d22ee4ec770cf8ea26301151ec6a8/datasets/coco.py#L33-L47

I think we should not add this to `util.misc` , but just modify the D2 wrapper code to do the conversions we need, by calling the necessary D2 functions there.

File: models/detr.py [hunk 3]
```diff
@@ -86,7 +86,7 @@ class SetCriterion(nn.Module):
         1) we compute hungarian assignment between ground truth boxes and the outputs of the model
         2) we supervise each pair of matched ground-truth / prediction (supervise class and box)
     """
-    def __init__(self, num_classes, matcher, weight_dict, eos_coef, losses):
+    def __init__(self, num_classes, matcher, weight_dict, eos_coef, losses, is_detectron=False):
```
Comment (fmassa) [inline models/detr.py:89]: ideally, there shouldn't be any mention of `detectron2` in the base implementation

**Comment by jd730** [comment]
Hi @fmassa, thank you for reviewing my PR. I will modify the code on August 15th. I am sorry about late reply and modification.

File: models/detr.py (hunk 1 above)
Comment (jd730) [reply models/detr.py:179]: Then, maybe making inheritance `SetCriterion` on `d2/detr` folder from `SetCriterion` on `models/detr.py` will be better.

Comment (fmassa) [reply models/detr.py:179]: I believe this should be changed in `prepare_targets` in https://github.com/facebookresearch/detr/blob/f4cdc542de34de771da8b9189742e5465f5220cd/d2/detr/detr.py#L160

You can add one extra line in the implementation in https://github.com/facebookresearch/detr/blob/f4cdc542de34de771da8b9189742e5465f5220cd/d2/detr/detr.py#L179-L188 to convert the polygons to bitmask

**Comment by jd730** [comment]
Hi @fmassa, I updated following your comments. Could you review it, please?

Thanks a lot.

File: d2/detr/detr.py [hunk 4]
```diff
@@ -111,6 +114,23 @@ def __init__(self, cfg):
...
+            frozen_weights = cfg.MODEL.DETR.FROZEN_WEIGHTS
+            if frozen_weights != '':
+                print("LOAD pre-trained weights")
+                weight = torch.load(frozen_weights, map_location=lambda storage, loc: storage)['model']
+                new_weight = {}
+                for k, v in weight.items():
+                    if 'detr.' in k:
+                        new_weight[k.replace('detr.', '')] = v
+                    else:
+                        print(k)
```
Comment (alcinos) [inline d2/detr/detr.py:127]:
```suggestion
                        print(f"Skipping loading weight {k} from frozen model")
```

**Comment by alcinos** [comment]
Looks good to me too, thanks! Merging.

**Comment by Real-YeJ** [comment]
@jd730 @fmassa  hey, thanks for great work!  I followed the steps to train a custom model of instance segmentation, and I used the convertered **detr-r50-e632da11.pth** as the input of **FROZEN_WEIGHTS**  

Here are the log of loss:
[10/22 19:20:10 d2.data.common]: Serializing 4809 elements to byte tensors and concatenating them all ...
[10/22 19:20:11 d2.data.common]: Serialized dataset takes 22.56 MiB
[10/22 19:20:11 d2.data.build]: Using training sampler TrainingSampler
[10/22 19:20:11 fvcore.common.checkpoint]: No checkpoint found. Initializing model from scratch
[10/22 19:20:11 d2.engine.train_loop]: Starting training from iteration 0
[10/22 19:20:18 d2.utils.events]:  eta: 5:36:07  iter: 19  total_loss: 867.5  loss_ce: 48.98  loss_bbox: 3.039  loss_giou: 2.704  loss_mask: 0.01238  loss_dice: 0.9895  loss_ce_0: 35.71  loss_bbox_0: 3.122  loss_giou_0: 2.586  loss_ce_1: 44.9  loss_bbox_1: 3.082  loss_giou_1: 2.531  loss_ce_2: 45.77  loss_bbox_2: 3  loss_giou_2: 2.615  loss_ce_3: 48.77  loss_bbox_3: 3.003  loss_giou_3: 2.669  loss_ce_4: 48.37  loss_bbox_4: 3.057  loss_giou_4: 2.717  time: 0.2763  data_time: 0.0351  lr: 0.0001  max_mem: 5358M

[10/22 19:20:24 d2.utils.events]:  eta: 5:43:54  iter: 39  total_loss: 903  loss_ce: 49.12  loss_bbox: 2.764  loss_giou: 2.384  loss_mask: 0.006315  loss_dice: 0.9862  loss_ce_0: 34.83  loss_bbox_0: 2.897  loss_giou_0: 2.372  loss_ce_1: 46.91  loss_bbox_1: 2.867  loss_giou_1: 2.405  loss_ce_2: 47.9  loss_bbox_2: 2.844  loss_giou_2: 2.451  loss_ce_3: 48.52  loss_bbox_3: 2.757  loss_giou_3: 2.406  loss_ce_4: 49.7  loss_bbox_4: 2.901  loss_giou_4: 2.4  time: 0.2946  data_time: 0.0027  lr: 0.0001  max_mem: 6277M

[10/22 19:20:30 d2.utils.events]:  eta: 5:47:56  iter: 59  total_loss: 863.5  loss_ce: 48.75  loss_bbox: 3.041  loss_giou: 3.101  loss_mask: 0.006423  loss_dice: 0.9822  loss_ce_0: 37.23  loss_bbox_0: 3.145  loss_giou_0: 2.965  loss_ce_1: 45.46  loss_bbox_1: 3.17  loss_giou_1: 2.968  loss_ce_2: 46.39  loss_bbox_2: 3.067  loss_giou_2: 2.95  loss_ce_3: 49.27  loss_bbox_3: 3.061  loss_giou_3: 3.033  loss_ce_4: 48.4  loss_bbox_4: 3.051  loss_giou_4: 3.092  time: 0.2911  data_time: 0.0019  lr: 0.0001  max_mem: 6277M

[10/22 19:20:35 d2.utils.events]:  eta: 5:39:11  iter: 79  total_loss: 855.7  loss_ce: 46.69  loss_bbox: 3.282  loss_giou: 2.721  loss_mask: 0.008903  loss_dice: 0.9588  loss_ce_0: 34.42  loss_bbox_0: 2.595  loss_giou_0: 2.585  loss_ce_1: 43.88  loss_bbox_1: 3.158  loss_giou_1: 2.72  loss_ce_2: 45.58  loss_bbox_2: 3.155  loss_giou_2: 2.671  loss_ce_3: 46.61  loss_bbox_3: 3.218  loss_giou_3: 2.701  loss_ce_4: 46.36  loss_bbox_4: 3.273  loss_giou_4: 2.67  time: 0.2834  data_time: 0.0018  lr: 0.0001  max_mem: 6277M

[10/22 19:20:41 d2.utils.events]:  eta: 5:39:06  iter: 99  total_loss: 888.1  loss_ce: 49.67  loss_bbox: 2.982  loss_giou: 2.661  loss_mask: 0.007313  loss_dice: 0.9738  loss_ce_0: 36.05  loss_bbox_0: 2.896  loss_giou_0: 2.573  loss_ce_1: 47.62  loss_bbox_1: 2.991  loss_giou_1: 2.695  loss_ce_2: 49.22  loss_bbox_2: 2.91  loss_giou_2: 2.68  loss_ce_3: 50.9  loss_bbox_3: 2.962  loss_giou_3: 2.623  loss_ce_4: 50.32  loss_bbox_4: 2.989  loss_giou_4: 2.638  time: 0.2836  data_time: 0.0018  lr: 0.0001  max_mem: 6277M

the total_loss is thousands of times as much as the loss_mask, which is so abnormal. Can you tell me some possible problems?

**Comment by jd730** [comment]
Hi @Real-YeJ,  Did you use `MODEL.WEIGHTS` at the same time?
Please refer to https://github.com/facebookresearch/detr/issues/252.

I think it might be good to make a new issue or make a comment on #252 rather than comment on PR.

**Comment by Real-YeJ** [comment]
@jd730 so sorry, I'll check the #252 .thanks!

**Comment by bowenc0221** [comment]
Hi @jd730, thanks for the effort. I'm wondering does the current D2 wrapper support inference on pre-trained DETR panoptic model? I tried to do inference but got 0 AP.

**Comment by jd730** [comment]
Hi @bowenc0221, I think the current version only supports instance segmentation but you may easily modify it for the evaluation.

**Comment by lxtGH** [comment]
@jd730 Hi! What are results of Instance Segmetnation using detr with this conifg ?
